Polishing the hardware documentation (#334)
XanaduAI/strawberryfields: doc/tutorials/tutorial_X8.rst

.. _starship:

Executing programs on X8 devices
================================

In this tutorial, we demonstrate how Strawberry Fields can execute

> quantum programs on the remote photonic X8 family of devices, via the Xanadu cloud platform. Topics

Reviewer: ```suggestion
quantum programs on the X8 family of remote photonic devices, via the Xanadu cloud platform. Topics
```

Reviewer: Just preferred that order, but either is fine

> .. _starship:

Reviewer 2: lol

.. _starship:

> Executing programs on X8 devices

Reviewer 2: Not a huge deal, but it still kinda reads a bit weird. This section refers to "X8 devices" and "X8 family", but everywhere else just refers to "X8 chip" etc. (singular).

Suggest moving the "What's in a name" callout to here (with the reduced size per my code suggestion)

Author: Have moved the reduced size callout up here :)

Author: I've also gone through and standardized any weird appearance of 'X8 chip'

===================================

In this tutorial, we demonstrate how Strawberry Fields can execute
quantum programs on the remote photonic X8 device, via the Xanadu cloud platform. Topics
covered include:

* Configuring Strawberry Fields to access the cloud platform,

> * Using the :class:`~.RemoteEngine` to execute jobs remotely on chip ``X8_01``,

Reviewer: One thing that came to my mind when I saw this:
"So is the chip `X8` or `X8_01`?"

Author: > explain the nomenclature of chips, i.e. X to encode architecture, 8 for number of spatial times frequency modes, _01 as a running number to identify different physical chips that are used in the hardware implementation, and finish with a comment that the specific chip instance must be chosen for running jobs on it.

I think this might be the best approach.

using the command line interface, see the :doc:`/introduction/photonic_hardware`
quickstart guide.

Device details
--------------

> Below, we will be submitting a job to run on the chip ``X8_01``. This is an 8 mode chip,

Reviewer 2: ```suggestion
Below, we will be submitting a job to run on the chip ``X8_01``. This is an 8-mode chip,
```

with the following restrictions:

* The initial states are two-mode squeezed states (:class:`~.S2gate`). We call modes 0 to 3 the

>   *signal modes* and modes 4 to 7 the *idler modes*. Two mode squeezing is between the pairs

Reviewer 2: ```suggestion
  *signal modes* and modes 4 to 7 the *idler modes*. Two-mode squeezing is between the pairs
```

  of modes: (0, 4), (1, 5), (2, 6), (3, 7).

> 

Reviewer 3: ```suggestion

At this point, only the parameters :code:`r=1`, :code:`phi=0` are allowed for the two-mode 
squeezing gates between any pair of signal and idler modes. Eventually, a range of 
squeezing amplitudes :code:`r` will be supported.   

```

      has 8 available modes.

    * ``A`` is the **architecture number.** This chip belongs to the ``X``-series of photonic
      devices. Devices in this series are characterized by the following properties:

      - The available modes :math:`M` are divided into two main groups, signal modes
        (numbered :math:`[0, M/2-1]`) and idler modes (numbered :math:`[M/2, M-1]`).

      - Initial two-mode squeezed states entangle the signal modes and the idler modes
        in a 'cascade'; entangled mode pairs are :math:`[(0, M/2), (1, M/2+1), \dots, (M/2-1, M-1)]`.

      - An arbitrary :math:`M/2\times M/2` unitary can then be applied to both the signal and idler modes.

      - Finally, the modes terminate with photon-number resolving detectors.

    * ``YY`` is the **chip ID**. It specifies a particular chip instance chosen for executing the jobs.
      While all chips in the same series have the same architecture, specific chips may be characterized
      by hardware imperfections, including loss and gate fidelity.

    While ``X8`` is used here to specify *any* chip in the ``X8``-series (including ``X8_01``), when

>     submitting remote jobs, a specific chip ID must be chosen.

Author: @lneuhaus and @trbromley, have added in a note box here!

Author: (I defined mode number before architecture number, because it is useful to have the number of modes already defined)

Reviewer: This is a cool box! I could see it being quite a popular bit of information.

One thing I wasn't sure on is whether we will really change from `X` when we go to other types of chip? E.g., the X series is strictly for two-mode squeezing chips? I'd potentially leave out the architecture number bullet point so as not to over specify at this stage. Although, if this rule has already been fixed, then I'm happy for it to go here.

Reviewer 3: > One thing I wasn't sure on is whether we will really change from X when we go to other types of chip? E.g., the X series is strictly for two-mode squeezing chips? I'd potentially leave out the architecture number bullet point so as not to over specify at this stage. Although, if this rule has already been fixed, then I'm happy for it to go here.

I agree this is not written in stone yet. Maybe @xanadu-zvernon has a more definite opinion on this?

Reviewer 3: This is correct as written above, so no need for change.

>     submitting remote jobs, a specific chip ID must be chosen.

Reviewer 3: ```suggestion
    submitting remote jobs, a reference to a specific chip (e.g.``X8_01``) must be chosen.
```

Device details
--------------

Below, we will be submitting a job to run on the chip ``X8_01``.

``X8_01`` is an 8-mode chip, with the following restrictions:

* The initial states are two-mode squeezed states (:class:`~.S2gate`). We call modes 0 to 3 the
  *signal modes* and modes 4 to 7 the *idler modes*. Two-mode squeezing is between the pairs
  of modes: (0, 4), (1, 5), (2, 6), (3, 7).

At this point, only the parameters :code:`r=1`, :code:`phi=0` are allowed for the two-mode
squeezing gates between any pair of signal and idler modes. Eventually, a range of

> squeezing amplitudes :code:`r` will be supported.

Reviewer: Might flow better below the two bullet points, i.e., below line 78.

Author: Done!

* Finally, the chip terminates with photon-number resolving measurements (:class:`~.MeasureFock`).

> .. figure:: /_static/X8.svg

Reviewer 2: something seems wrong with the line about SU(4) at the bottom of the image

Author: The sentence "Programmable SU(4) transformation"? I can change that to "Arbitrary 4x4 unitary"(?)

This comes straight from the hardware documentation, I'm also worried the single mode squeezers followed by entanglers in the circuit diagram might be confusing; best to change this to two mode squeeze gates?

Author: I might merge this in now, and we can update the image in a new PR :)

>>> counts[(0, 0, 0, 0, 0, 0, 0)]
2

Job management
~~~~~~~~~~~~~~

Above, when we called ``eng.run()``, we had to wait for the
remote device to execute the program and the result to be returned before we could
continue executing code. That is, ``eng.run()`` is a **blocking** method.

Sometimes, however, it is useful to submit the program
and continue performing computation locally, every now and again checking to see
if the job is complete and the results are ready. This is possible with
the **non-blocking** :meth:`eng.run_async() <.RemoteEngine.run_async>` method:

>>> job = engine.run_async(program, shots=100)

This method is complete as soon as the job has been submitted; no need to wait for

> the job to move through the queue, execute, and for the results to be returned.

Reviewer 2: A bit of confusing phrasing. You still need to wait for these things before you see your results, you're just not blocked at the prompt until then

Reviewer: Yes, maybe just remove this sentence. I think it's clear what's going on

> the job to move through the queue, execute, and for the results to be returned.

Author: ```suggestion
```

Unlike ``eng.run()``, it returns a :class:`~.Job` instance, which allows us to
check the status of our submitted job:

>>> job.id
"e6ead866-04c9-4d48-ba28-680e8639fc41"
>>> job.status
"queued"

When the job result is ready, it is available via the ``result`` property.
If the job result is not yet available, however, an ``InvalidJobOperationError``
will be raised:

>>> job.result
InvalidJobOperationError

To check when the results are ready, the job can be refreshed, and the status
checked:

>>> job.refresh()
>>> job.status
"complete"
>>> result = job.result
>>> result.samples.shape
(100, 8)

Finally, an incomplete job can be *cancelled* by calling :meth:`job.cancel() <.Job.cancel>`.

.. note::

>     The :class:`~.operation` decorator allows you to create your own Strawberry Fields

Reviewer: I think this note has got left behind and belongs more naturally in the previous section.
E.g., I'd put it around line 130

Author: Have moved it up!

>>> counts[(0, 0, 0, 0, 0, 0, 0)]
2

.. note::

    The :class:`~.operation` decorator allows you to create your own Strawberry Fields
    operation. This can make it easier to ensure that the same unitary is always
    applied to the signal and idler modes.

    .. code-block:: python3

        from strawberryfields.utils import operation

>         @operation(4)

Reviewer 2: I asked this before, but I might have missed the answer. Is it not possible to make `q` an argument to the decorator rather than the qfunc?

Author: See discussion here: https://github.com/XanaduAI/strawberryfields/pull/330#discussion_r396824959

  This performs a rectangular decomposition using Mach-Zehnder interferometers.

* You can use :class:`~.ops.BipartiteGraphEmbed` to embed a bipartite graph on
  the photonic device.

  .. warning::

      Decomposed squeezing values depend on the graph
      structure, so only bipartite graphs that result in equal squeezing on all

>       modes can be executed on ``X8_01`` chips.

Reviewer 2: ```suggestion
      modes can be executed on ``X8`` chips.
```

Reviewer: I feel like we're not clear on devices vs chips and X8 vs X8_01.
Is X8 a family of 8-mode chips (maybe this is akin to a "device") and X8_01 a particular chip?

Author: Yep, exactly! I will add a paragraph explaining this.

current working directory. You can also omit the ``--output`` parameter to print the
result to the screen.

.. note::

    Saved Blackbird scripts can be imported as Strawberry Fields programs
    using the :func:`~.load` function:

    >>> prog = load("test.xbb")

    Strawberry Fields programs can also be exported as Blackbird scripts
    using :func:`~.save`:

    >>> sf.save("program1.xbb", prog)

Embedding bipartite graphs
--------------------------

> The X-series of devices support embedding bipartite graphs,

Reviewer 2: ```suggestion
The X8 device class supports embedding bipartite graphs,
```

    chip structure and architecture.

    At Xanadu, photonic devices are named using the nomenclature ``AM_YY``, where:

    * ``M`` is the **mode number**, the number of modes available on the chip. ``X8_01``
      has 8 available modes.

    * ``A`` is the **architecture number.** This chip belongs to the ``X``-series of photonic
      devices. Devices in this series are characterized by the following properties:

      - The available modes :math:`M` are divided into two main groups, signal modes
        (numbered :math:`[0, M/2-1]`) and idler modes (numbered :math:`[M/2, M-1]`).

      - Initial two-mode squeezed states entangle the signal modes and the idler modes
        in a 'cascade'; entangled mode pairs are :math:`[(0, M/2), (1, M/2+1), \dots, (M/2-1, M-1)]`.

      - An arbitrary :math:`M/2\times M/2` unitary can then be applied to both the signal and idler modes.

      - Finally, the modes terminate with photon-number resolving detectors.

    * ``YY`` is the **chip ID**. It specifies a particular chip instance chosen for executing the jobs.
      While all chips in the same series have the same architecture, specific chips may be characterized

>       by hardware imperfections, including loss and gate fidelity.

Reviewer 2: ```suggestion
```

    * ``A`` is the **architecture number.** This chip belongs to the ``X``-series of photonic
      devices. Devices in this series are characterized by the following properties:

      - The available modes :math:`M` are divided into two main groups, signal modes
        (numbered :math:`[0, M/2-1]`) and idler modes (numbered :math:`[M/2, M-1]`).

      - Initial two-mode squeezed states entangle the signal modes and the idler modes
        in a 'cascade'; entangled mode pairs are :math:`[(0, M/2), (1, M/2+1), \dots, (M/2-1, M-1)]`.

      - An arbitrary :math:`M/2\times M/2` unitary can then be applied to both the signal and idler modes.

      - Finally, the modes terminate with photon-number resolving detectors.

    * ``YY`` is the **chip ID**. It specifies a particular chip instance chosen for executing the jobs.
      While all chips in the same series have the same architecture, specific chips may be characterized
      by hardware imperfections, including loss and gate fidelity.

    While ``X8`` is used here to specify *any* chip in the ``X8``-series (including ``X8_01``),
    you may also specify a specific chip ID (such as ``X8_01``) during program compilation

>     and execution.

Reviewer: Is it worth thinking about not having this part at all?

> ``X8`` is an 8-mode chip, with the following restrictions:

This sentence from further up above, the following paragraph, and image I think are enough for the user to know what's going on at this point.

I know this part might have taken some time to write up, but maybe it can be reused in a future iteration.

Author: I agree, it isn't necessary. It would be nice to find a home for this note box though, as it is useful info for research (i.e, you want to guarantee you are submitting to the same chip in case defaults change, for reproducibility).

Reviewer 2: Yep, please remove for now from this page

Reviewer 2: Note also that there's a difference in explaining the physical chip architecture and explaining the naming scheme. First one is fine to have on this page, but second one not necessary
